File: java/148_SortList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        int length = 1;
        ListNode curr = head;
        while (curr != null) {
            curr = curr.next;
            length++;
        }
        return splitAndMerge(head, length);
    }
    
    public ListNode splitAndMerge(ListNode head, int length) {
        if (head == null || length == 1) return head;
        
        int half = length / 2, idx = 0;
        ListNode curr = head, prev = null;
        while (curr != null && idx < half) {
            prev = curr;
            curr = curr.next;
            idx++;
        }
        
        if (prev != null) prev.next = null;
        head = splitAndMerge(head, half);
        curr = splitAndMerge(curr, half % 2 == 1 ? half + 1 : half);
        
        return mergeTwoLists(head, curr);
    }
    
    public ListNode mergeTwoLists(ListNode firstList, ListNode secondList) {
        if (firstList == null || secondList == null) return firstList != null ? firstList : secondList;
        else if (firstList.val > secondList.val) return mergeTwoLists(secondList, firstList);
        
        firstList.next = mergeTwoLists(firstList.next, secondList);
        return firstList;
    }
    
}

File: swift/148_SortList.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func sortList(_ head: ListNode?) -> ListNode? {
        // get initial length
        var length = 0
        var curr = head
        while curr != nil {
            curr = curr?.next
            length += 1
        }
        
        return splitAndMergeLists(head, length)
    }
    
    func splitAndMergeLists(_ head: ListNode?, _ length: Int) -> ListNode? {
            if head == nil || length == 1 { return head }
            var curr = head, prev: ListNode? = nil
            var half = length / 2, index = 0
            while curr != nil && index < half {
                prev = curr
                curr = curr?.next
                index += 1
            }

            if prev != nil { prev?.next = nil }
            var node = splitAndMergeLists(head, half)
            curr = splitAndMergeLists(curr, half % 2 == 1 ? half + 1 : half)
            return mergeTwoLists(node, curr)
        }
        
        func mergeTwoLists(_ smallerHead: ListNode?, _ biggerHead: ListNode?) -> ListNode? {
            if smallerHead == nil || biggerHead == nil { return smallerHead != nil ? smallerHead : biggerHead }
            else if smallerHead!.val > biggerHead!.val { return mergeTwoLists(biggerHead, smallerHead) }
            
            var smaller = smallerHead
            smaller?.next = mergeTwoLists(smaller?.next, biggerHead)
            return smaller
        }
}
